source/Condition.h

```diff
@@ -0,0 +1,340 @@
+
+#include "ConditionsStore.h"
+
+
+
+// This class stores either:
+//   1. A condition that is in a ConditionsStore
+//   2. A literal value (key is empty)
+//
+// The ValueType (V in the template) is first in the template since
+// (nearly?)  all Condition classes will use a std::string key (K). It
+// should be an arithmetic type, like double, int64_t, int, or
+// unsigned. A bool should work too, but that's untested.
+
+template < class V >
+class Condition {
+	template <class V2>
+	friend class Condition;
+public:
+	using ValueType = V;
+	using KeyType = std::string;
+	using StoreType = ConditionsStore;
+
+	static_assert(std::is_arithmetic<ValueType>::value, "Condition value type must be arithmetic.");
+	static_assert(std::is_class<KeyType>::value, "Condition key type must be a class.");
+	static_assert(&KeyType::empty, "Condition key must have an empty function.");
+
+	Condition() : value(), key() {}
+	explicit Condition(const V &value) : value(value), key() {}
+
+	// Initialize the condition with the specified value.
+	// Record what store and key to use, but do not query the store for a value.
+	Condition(const V &value, std::shared_ptr<ConditionsStore> store, const KeyType &key);
+
+	// Get the initial value from the store. Use V() if the store returned nothing.
+	Condition(std::shared_ptr<ConditionsStore> store, const KeyType &key);
+
+	template <class V2>
+	Condition(const Condition<V2> &other);
+
+	template <class V2>
+	Condition &operator=(const Condition<V2> &other);
+
+	// This commented-out operator may look convenient, but it causes
+	// everything to break because any Condition=Condition assignment uses
+	// this operator instead of the Condition=Condition implementation,
+	// preventing explicit conversion of Conditions. A lot of code would
+	// be simpler if this was working.
+	//
+	// template <class T, typename std::enable_if<std::is_arithmetic<T>::value>::type* = nullptr>
+	// Condition &operator=(const T &other);
+
+	// Send the latest value to the ConditionsStore
+	const ValueType &SendValue();
+
+	// Read the value from the ConditionsStore
+	const ValueType &UpdateConditions();
+
+	// Read the value from the ConditionsStore, pass it through a
+	// filter, and use the new result. The filter arguments are:
+	// 1. The ConditionsStore's value, as a ConditionsStore::ValueType
+	// 2. The current value in the Condition, as a Condition::ValueType
+	// The filter should return the new value as a Condition::ValueType
+	template <class Filterer>
+	const ValueType &UpdateConditions(Filterer validator);
+
+	// Pass the current value through a filter. The filter
+	// receives the current value as a ValueType and returns the
+	// new value as a ValueType
+	template <class Filterer>
+	const ValueType &Filter(Filterer filter);
+
+	// Accessors and mutators
+
+	const ValueType &Value() const { return value; }
+	ValueType &Value() { return value; }
+	const KeyType &Key() const { return key; }
+	KeyType &Key() { return key; }
```

Comment:
can we move these to the cpp?
if so do put whatever you can there
if not I think we should follow existing conventions and declare these inline at the end of the class (like for damagedealt.h)

Reply:
This matches what is done in the other template classes.

Reply:
It can't go in the cpp because it is a template class. We could add .tpp files, like some projects do, which move the implementation of the template class to another file. Every file that include Condition.h would still need Condition.tpp, so the Condition.h would have an `#include "Condition.tpp"` at the end.

Reply:
lets not complexify stuff with .tpp, this is ok